src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/Checksum.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.AspNetCore.Razor.Utilities;
+
+internal sealed partial class Checksum : IEquatable<Checksum?>
+{
+    private const int HashSize = 20;
+
+    public static readonly Checksum Null = new(default);
+
+    public readonly HashData Data;
+
+    public Checksum(HashData data)
+        => Data = data;
+
+    public static Checksum From(ReadOnlySpan<byte> source)
+    {
+        if (source.Length == 0)
+        {
+            return Null;
+        }
+
+        if (source.Length < HashSize)
+        {
+            throw new ArgumentException($"{nameof(source)} must be equal to or larger than the hash size: {HashSize}", nameof(source));
+        }
+
+        if (!MemoryMarshal.TryRead(source, out HashData hash))
+        {
+            throw new InvalidOperationException("Could not read hash data");
+        }
+
+        return new Checksum(hash);
+    }
+
+    private void WriteTo(BinaryWriter writer)
+        => Data.WriteTo(writer);
+
+    public string ToBase64String()
+    {
+#if NETCOREAPP
+        Span<byte> bytes = stackalloc byte[HashSize];
+        MemoryMarshal.TryWrite(bytes, ref Unsafe.AsRef(in Data));
+
+        return Convert.ToBase64String(bytes);
+#else
+        unsafe
+        {
+            var data = new byte[HashSize];
+            fixed (byte* dataPtr = data)
+            {
+                *(HashData*)dataPtr = Data;
+            }
+
+            return Convert.ToBase64String(data, offset: 0, length: HashSize);
+        }
+#endif
+    }
+
+    public static Checksum FromBase64String(string value)
+        => value is null
+            ? Null
+            : From(Convert.FromBase64String(value));
+
+    public override bool Equals(object? obj)
+        => Equals(obj as Checksum);
+
+    public bool Equals(Checksum? other)
+        => other is not null &&
+           Data == other.Data;
+
+    public override int GetHashCode()
+        => Data.GetHashCode();
+
+    public override string ToString()
+        => ToBase64String();
+
+    public static bool operator ==(Checksum? left, Checksum? right)
```

Comment:
Is there a reason this couldn't be a record? Seems like most of this equality could be auto-implemented by that.

Reply:
You're totally right, and I'll make it a record. I had mostly copied this from Roslyn, where it was already in non-record form.